### DISCUSSION TITLE
Docker runtime Environment Variables doesn't work

### DISCUSSION THREAD

**Discussion by gyto23** [vercel/next.js#15651] [Help]
So I spend a lot of time debugging this and trying to figure out a better approach to build application on the docker. After running a simple `Dockerfile` I have mention that non of my .env variables doesn't work, and I assume this is really common problem after 9.4 release where we have to specify env var as `NEXT_PUBLIC_` (which doesnt work for on runtime configurations) I know that I can go with a `ARG` on buildtime but this is not a option for me because the application is multi domain application which needs different env var in different domains. My general question to the community of next.js is there a way and how to trigger runtime variables after application image is build and before docker container is running? 

P.S. next.js application uses a `express` server as custom server, and I dont use any of the `serverless` features 

```
FROM node:alpine

ENV TEMP_DIR /app

# Create app directory
RUN mkdir -p ${TEMP_DIR}
WORKDIR ${TEMP_DIR}
RUN apk add --no-cache libc6-compat

ENV NODE_ENV production
ENV PORT 3000

EXPOSE 3000

COPY .npmrc ${TEMP_DIR}
COPY package.json ${TEMP_DIR}
COPY yarn.lock ${TEMP_DIR}
COPY faviconDist.json ${TEMP_DIR}

RUN yarn install --no-optional
RUN rm -f .npmrc

COPY . ${TEMP_DIR}

RUN yarn build

RUN addgroup -g 1001 -S nodejs
RUN adduser -S nextjs -u 1001

USER nextjs

CMD ["yarn", "start"]

```

**Comment by jamesmosier** [comment]
If you don't want to have your environment variables built into your bundles, you could take a look at the [runtime configuration](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) options.

As mentioned in this doc though:

> Generally you'll want to use build-time environment variables to provide your configuration. The reason for this is that runtime configuration adds rendering / initialization overhead and is incompatible with Automatic Static Optimization.

**Reply by gyto23** [reply]
@jamesmosier they don't work either, I have tried to use them too with docker, I assume this is nextjs issue or something because they don't pop up in the client side, but they are on server side

**Reply by jamesmosier** [reply]
Could you create a repository that have a basic setup that would help others track down this issue and help? 

Creating a `next.config.js` with this:

```js
module.exports = {
  publicRuntimeConfig: {
    myPublicVar: 'this is public!',
  },
}

```

and then in my page I have:

```jsx
import getConfig from 'next/config';
const { publicRuntimeConfig } = getConfig();

export default function Home() {
  return <div>{publicRuntimeConfig.myPublicVar}</div>
}
```

results in `<div>this is public!'</div>` being rendered when I run my Docker container.

**Reply by gyto23** [reply]
Sure give me some time for it

**Reply by gyto23** [reply]
@jamesmosier actually just tried on my setup (I am running late on everything so I did not have to create example, but will do it soon, sorry) Anyway I tested in a way that next.config.js is setup as

```
module.exports = {
  publicRuntimeConfig: {
    myPublicVar: process.env.FANCY_ENV_VAR,
  },
}
```

looks like next.config.js dont have access to the environments like this on runtime because I need to access a sensitive variables that should not be on client side when they open the webpage. On the other hand, your setup works fine which is not what I am actually expecting from app.

**Reply by mAAdhaTTah** [reply]
@gyto23 Part of the reason you may be experiencing inconsistent results is pages that are statically-generated can't take advantage of the runtime config. I'm in a similar situation as you, and unfortunately, it's not clear I can take advantage of statically generated pages until the static generation is pushed off to runtime.

**Reply by gyto23** [reply]
@mAAdhaTTah which super odd, right? I am trying to build a nextjs project using a docker compose option where variables actually running of the scripts that I wrote to deploy application to the cloud. In addition, it makes harder every time to remember which script should go and what env var have to be used in, that why I came up with idea to use runtime var to populate them in the script after the build is done

**Reply by mAAdhaTTah** [reply]
@gyto23 Yeah... we ended up using a bit of a hack to ensure none of the static HTML gets built during the build and it's all generated at runtime. We did this by using `gSP` but returning an empty array during build for the `paths` and using `fallback: 'unstable_blocking'`, then when the first request comes in, we return the _real_ array of paths, so the HTML isn't generated until the first request, which allows it to use the runtime cars.

I'd prefer this to be a "blessed" implementation, but given Vercel itself doesn't run on Docker, I'm not sure they're planning on making Docker support easier.

**Reply by gyto23** [reply]
@mAAdhaTTah its really interesting hack you did. However, I prefer not to implement any weird hacks on the platform I am developing in. I have battle with nextJs for months, and only way I come up to avoid further issues is to use `getServerSideProps` and pass the keys I need to the higher components, regarding the runtime env var is still big question to us how to solved, even after using fancy way export on `publicRuntimeConfig` because it might only work vercel platform, Docker ignoring it for some reason

**Comment by Manikanta-20** [comment]
@jamesmosier 
how to use docker runtime environment variables and already  i'm using .env with custom server(Express.Js) .......can we use both environment /fetch docker environment variables at run time?

**Comment by kyma** [comment]
I also have this problem. publicRuntimeConfig doesn't pick up  process.env.ANYTHING. Is there any solution to this?

**Reply by gyto23** [reply]
@kyma per my research still not, waiting for vercel team to this treat and read about issues my team and I have, hope they going to see it one day

**Reply by reanimatedmanx** [reply]
https://github.com/vercel/next.js/discussions/17641#discussioncomment-110484

Guys I found the solution, basically if you don't really need https://nextjs.org/docs/advanced-features/automatic-static-optimization

Then just add an empty 

```js
//pages/_app.js|.tsx
MyApp.getInitialProps = async (appContext: AppContext) => {
  const appProps = await App.getInitialProps(appContext)

  return { ...appProps }
}
```

Then you will be able to access publicRuntimeConfig env variables from client in runtime even on your Dockerized app on stage/prod

**Reply by gunar** [reply]
NB: `App` comes from here:

```ts
import App from 'next/app'
```

**Comment by sethsnel** [comment]
I seem to have a similar issue. I'm still running on Nextjs `9.5.2`.

I try to load my runtime config in `next.config.js` like this:
<img width="365" alt="image" src="https://user-images.githubusercontent.com/3225809/125818659-192f1575-a00a-4ec2-ad31-299004d7d7da.png">

My dockerfile looks like this:
<img width="490" alt="image" src="https://user-images.githubusercontent.com/3225809/125818855-8849f5a1-10d9-4dcb-9822-754e36a88715.png">

And when I run the image with custom env I have logged output:
`docker run -p 3000:3000 -d --name notify-app -e AUTH_SERVER_ENDPOINT=temp -e HASURA_ENDPOINT=temp ghcr.io/sethsnel/notify-app`
<img width="371" alt="image" src="https://user-images.githubusercontent.com/3225809/125818954-328d3814-dbab-4c00-bae7-d45741ab6626.png">

However, in my code I get the env values which where defined during build time:
<img width="392" alt="image" src="https://user-images.githubusercontent.com/3225809/125819120-3cbfcb36-bd02-4834-9117-884dda53f358.png">

And yes I do use `getInitialProps` on my page.
My use case is that I want to verify my image on acceptance environment with integration tests.
If successful then the same image will be used to deploy to prd, where obviously the environment variable will be different.

Hence I do not want to set the environment variables fixed when I'm building the image.
But I want my app to used the appropriate assigned runtime EV.

Does anyone have a lead on what I'm overlooking?

**Reply by sethsnel** [reply]
@timneutkens Is this a feature or a bug?

**Reply by gyto23** [reply]
@timneutkens More like a bug, but facing the same issue I still have, and it was a year ago :)

**Reply by AnthonyDugarte** [reply]
Any updates on this? 

I'm using the suggested [dockerfile](https://nextjs.org/docs/deployment#docker-image) but runtime configs are not working.

**Reply by Eitan1112** [reply]
I found that using the suggested dockerfile won't work, replace it with the following:
The main changes are that we copy all the build folder from the builder stage instead of only the js files, and then start the container with the yarn start command, which makes it work. Probably something in the optimization process there messes it up, not really sure, but that's a good workaround 


```
# Install dependencies only when needed
FROM node:16-alpine AS deps
# Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
RUN apk add --no-cache libc6-compat
WORKDIR /app
COPY package.json yarn.lock ./
RUN yarn install --frozen-lockfile

# Rebuild the source code only when needed
FROM node:16-alpine AS builder
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY . .

ENV NEXT_TELEMETRY_DISABLED 1
ENV DOCKER_BUILD 1

RUN yarn build


# Production image, copy all the files and run next
FROM node:16-alpine AS runner
WORKDIR /app

ENV NODE_ENV production
# Uncomment the following line in case you want to disable telemetry during runtime.
ENV NEXT_TELEMETRY_DISABLED 1

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

COPY --from=builder /app/ ./

USER nextjs

EXPOSE 3000

ENV PORT 3000

CMD ["yarn", "start"]

```
